### PR TITLE
Upgrade to Bootstrap 4

### LINKED ISSUE #1716: Upgrade to Bootstrap 4
We need to upgrade to Bootstrap 4, and this issue will hold a checklist for the same in the future. For now, we need to consolidate all the tasks we need to perform the migration and make a list of them in here, then break them into small self-contained and easy to solve issues.

Reference:
1. http://getbootstrap.com/docs/4.0/migration/
2. http://www.dotnetcurry.com/javascript/1334/migrating-bootstrap-3-to-bootstrap-4
3. http://upgrade-bootstrap.bootply.com (you can put HTML in here and it'll help you convert it!) 

/cc @jywarren 

# Checklist

## General Recommendations / Miscellaneous

- [x] Make a branch called something along the lines of `bootstrap-4` in the `publiclab/plots2` repository. All PRs made for this issue and all related ones need to use that branch as a base and not `master`. That branch would be merged into `master` only when the migration would be complete, we would verify the results and optionally, bootstrap 4 comes out of beta.
- [ ] Fixing pagination
- [x] Fixing the navigation bar

## Deprecated --- Replace with modern alternatives

- [x] Refactor all existing `well`s into `card`s.
- [ ] Refactor all existing `panel`s into `card`s.
- [ ] Refactor all existing `thumbnail`s into `card`s.
- [x] Glyphicons, if any, should be replaced by other icons. Let's use only font-awesome for all our icons, they are good enough for most of our use cases.

## Grid System

- [ ] Bump all grid sizes by one. (eg: `sm` becomes `md`, `md` becomes `lg` and so on).
- [ ] Add support for new `xs` tier for improving support at smaller mobile/IoT devices.

### PR DESCRIPTION
**Pull request by Souravirus**
Fixes #1716

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [ ] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `rake test`
* [X] code is in uniquely-named feature branch and has no merge conflicts
* [X] PR is descriptively titled
* [ ] ask `@publiclab/reviewers` for help, in a comment below

> We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged** if your tests fail at first!

If tests do fail, click on the red `X` to learn why by reading the logs.

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/contributing-to-public-lab-software 

Thanks!

### CODE REVIEW

**Comment by jywarren** [comment]
@edieblu take a look - @Souravirus is working on a Bootstrap 4 upgrade :-)

**Comment by jywarren** [comment]
Not sure why Travis is failing. We are seeing some intermittent issues with Dangerbot in #3935 that we are looking to resolve, in the meantime we can restart travis from time to time 👍

**Comment by jywarren** [comment]
OH! haha - this is actually a JS failure!

```

Error: Bootstrap tooltips require Tether (http://tether.io/)
  # bootstrap/dist/js/bootstrap.min.self-f0ac71a63207973b5b1ad57e55be7fdbf03e684db1ec6dfeccc294db81110ead.js:8
  # bootstrap/dist/js/bootstrap.min.self-f0ac71a63207973b5b1ad57e55be7fdbf03e684db1ec6dfeccc294db81110ead.js:8
TypeError: undefined is not a function (evaluating '$("[rel=popover]").popover({
    container: 'body',
    trigger: 'focus click'
  })')
  # setup.self-7a311c224e5c8bcfa91982d4a67f4633ac84e5da786cb857beb9812f1ec6e073.js:6
  # jquery.self-bd7ddd393353a8d2480a622e80342adf488fb6006d667e8b42e4c0073393abee.js:3233 -- fire
  # jquery.self-bd7ddd393353a8d2480a622e80342adf488fb6006d667e8b42e4c0073393abee.js:3363 -- fireWith
  # jquery.self-bd7ddd393353a8d2480a622e80342adf488fb6006d667e8b42e4c0073393abee.js:3583 -- ready
  # jquery.self-bd7ddd393353a8d2480a622e80342adf488fb6006d667e8b42e4c0073393abee.js:3618 -- completed
.FF...F
Failures:
  1) Plots2 sends a like request when like button is clicked
     Failure/Error: expected '0' to deeply equal '4'
  2) Plots2 unlikes a request if already liked
     Failure/Error: Attempted to wrap ajax which is already wrapped
  3) Wikis adds table CSS
     Failure/Error: undefined is not a function (evaluating 'element.find("[rel=tooltip]").tooltip()')
Finished in 0.07200 seconds
7 examples, 3 failures
Failed examples:
teaspoon -s default --filter="Plots2 sends a like request when like button is clicked"
teaspoon -s default --filter="Plots2 unlikes a request if already liked"
teaspoon -s default --filter="Wikis adds table CSS"
```

**Comment by Souravirus** [comment]
Yeah right @jywarren

**Comment by Souravirus** [comment]
Hi @jywarren, can you help me with the installation of bootstrap v4. I have tried updating the bootstrap package in yarn but it seems that bootstrap 4 is not working with the html elements. Thank you!!

**Comment by jywarren** [comment]
Ah, what exactly are the errors? Perhaps @ryzokuken has some thoughts to
contribute? Thanks!!

**Comment by Souravirus** [comment]
The thing is that I am writing bootstrap 4 classes but it is having no effect on the HTML elements. It is appearing as there is no bootstrap installed.

**Comment by Souravirus** [comment]
I think its working now. Thank you!!

**Comment by edieblu** [comment]
Hi @Souravirus nice to meet you. I'm working on a welcome landing page that is using Bootstrap 4, so really looking forward to your upgrade! :) How is it going so far?

**Comment by Souravirus** [comment]
Hi @edieblu nice to meet you. I am currently doing with the setup of bootstrap as the classes are not working well with the public lab. Just playing with Gemfile and yarn to set the correct configuration of bootstrap 4.

**Comment by jywarren** [comment]
This is awesome. Any screenshots to share? Keep up the good work - glad you got past that issue!

**Comment by Souravirus** [comment]
Hi @jywarren currently I am busy with my tests and assignments in my college. So, I am not able to work on it. I will surely work on with it once I get time. I don't think bootstrap is working fine still. As per I have studied, I have to transfer the stylesheet assets from .css to .scss. I guess it will work fine then. Till then I can show you the screenshots of the current homepage and how it looks. 

screenshot from 2018-11-14 20-59-02

**Comment by jywarren** [comment]
oh, i'm not sure, but I would be surprised if we need to convert all .css
to .scss... i would guess we need to change the classnames as shown in the
HTML in /app/views/ to match the new Bootstrap 4 classnames... you're going
down the checklist provided by @ryzokuken in the original issue?

**Comment by jywarren** [comment]
hmm, can you push it to unstable so I can take a look at it?

**Comment by Souravirus** [comment]
Hi @jywarren, I wanted you to show the changes in the unstable branch but there is an error coming with the yarn installation. Here is the error:
Integrity check failed for "bootstrap" (computed integrity doesn't match our records, got "sha1-ITs38caqGm3TiqX7/Mae+fuHCdo=")
Here is the link of the jenkins test:
https://jenkins.laboratoriopublico.org/job/Plots-Unstable/334/console

So now, the thing we can do is to install bootstrap 4 through gems. What do you think about this?

**Comment by jywarren** [comment]
Hi, i'm seeing a similar view to your screenshots above. I think the navbar structure will need to be reorganized in the new bootstrap 4 html structure. But I don't think anything is too far wrong here, besides the need to go through the checklist and upgrade all the css classes... I'd work on the navbar first as you've started to. I think there may be structural changes to the HTML as well to read through the Bootstrap docs and upgrade notes on. Once you complete the navbar, other areas will start to be easier too. See for example /wiki doesn't look too bad.

**Comment by Souravirus** [comment]
Hi, it's really nice to see that you jumped in. So, I will try to point out where I have changed some of the classes but the bootstrap simply didn't respond.

File: app/views/layouts/_header.html.erb [hunk 1]
```diff
@@ -7,15 +7,12 @@
   <%= stylesheet_link_tag "I18n" %>
 <% end %>
 
-<nav id="header" class="navbar navbar-inverse navbar-fixed-top">
-  <div class="container">
+<nav id="header" class="navbar navbar-expand-sm navbar-inverse bg-inverse">
```
Comment (Souravirus) [inline app/views/layouts/_header.html.erb]: Here, navbar-expand-sm should have expanded the navbar desktop view. But the navbar remained collapsed.

File: app/views/layouts/_header.html.erb [hunk 2]
```diff
@@ -7,15 +7,12 @@
...
     <!-- ||| menu, displayed for mobile: -->
-    <div class="navbar-header">
-      <a class="navbar-brand" id="brand" href="<%= root_path %>">Public Lab</a>
-      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#header-navbar-collapse">
-        <span class="fa fa-white fa-bars"></span>
-      </button>
-    </div>
+  <a class="navbar-brand" id="brand" href="<%= root_path %>">Public Lab</a>
+  <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#header-navbar-collapse">
```
Comment (Souravirus) [inline app/views/layouts/_header.html.erb]: This is the button which should have disappeared in large screens, but it didn't.

**Comment by jywarren** [comment]
I have kind of limited time today unfortunately -- have you tried feeding the navbar code through:

http://upgrade-bootstrap.bootply.com as mentioned in https://github.com/publiclab/plots2/issues/1716

Also have you checked in the dev console if the styles are actually being applied? 

I'm reading carefully here: https://getbootstrap.com/docs/4.1/components/navbar/

and here: http://getbootstrap.com/docs/4.0/migration/#navbar

File: app/views/layouts/_header.html.erb (hunk 1 above)
Comment (jywarren) [reply app/views/layouts/_header.html.erb]: Looking... i see `Navbars are responsive by default, but you can easily modify them to change that. Responsive behavior depends on our Collapse JavaScript plugin.` -- do we need that plugin added?

https://getbootstrap.com/docs/4.1/components/navbar/

Comment (jywarren) [reply app/views/layouts/_header.html.erb]: yes maybe, it also says `.navbar-toggler for use with our collapse plugin and other navigation toggling behaviors.`

Comment (jywarren) [reply app/views/layouts/_header.html.erb]: is popper also needed? not sure for what... https://getbootstrap.com/docs/4.1/getting-started/introduction/#js

Comment (jywarren) [reply app/views/layouts/_header.html.erb]: https://getbootstrap.com/docs/4.1/components/collapse/#how-it-works but i can't see where the plugin is included...

Comment (jywarren) [reply app/views/layouts/_header.html.erb]: hmm but this says it's all compiled in: https://getbootstrap.com/docs/4.1/getting-started/javascript/#individual-or-compiled

Comment (jywarren) [reply app/views/layouts/_header.html.erb]: GOT IT! 

https://github.com/publiclab/plots2/blob/master/app/assets/stylesheets/style.css#L116-L154

Overrides over the years as we've tried to customize and tweak. Just be sure we respect the behaviors shown here, or something equivalent! One complaint with the previous navbar was that it was too eager to hide things as the screen got narrower. This gif shows exactly how it was solved; hopefully the new Bootstrap just makes this much easier!! 

https://github.com/publiclab/plots2/pull/3491

**Comment by Souravirus** [comment]
Hi @jywarren, I am currently seeing the navbar changes caused by the style.css. Maybe it makes the navbar looks good. Thanks for your investigation into this.

**Comment by Souravirus** [comment]
I tried many things but the babbar is not cahnging in any way I dont know why. Please take a look on it.

**Comment by jywarren** [comment]
Hmm, interesting. We should be able to continue using Sprockets to provide direct access to yarn-included dependencies. But i'm not familiar with webpacker, but is it trying to replace sprockets (which is pretty old?).

**Comment by jywarren** [comment]
Did you want to push a new commit so i can see?

**Comment by Souravirus** [comment]
Hi @jywarren, sorry for the delay as I was stuck with the practical exams. But yeah I can work today for some time. Will be sending a commit as soon as possible by removing the conflicting styles. Thank you very much!!

**Comment by Souravirus** [comment]
Here is the screenshot after commenting those lines in style.css. Doesn't looks much great.
screenshot from 2018-11-21 22-31-28

**Comment by Souravirus** [comment]
Hi @jywarren, finally the navbar is looking a little bit decent. This worked by upgrading to bootstrap 4.0.0.
Previously I was using bootstrap 4.0.0-alpha. So, I guess that was causing the problems. Here are two screenshots of how the navbar is currently looking.
screenshot from 2018-11-22 02-44-25
screenshot from 2018-11-22 02-45-08

**Comment by jywarren** [comment]
Oooh! Getting closer! Thanks, Sourav - great work. Hope your exams went
well!

**Comment by Souravirus** [comment]
Yeah exams are going great. Thank you!!

**Comment by jywarren** [comment]
Hi, just checking in here! How are things going? Thanks!!!

**Comment by Souravirus** [comment]
Hi @jywarren, I will resume this work from 8th of December as there are exams going on currently in our college. I am really sorry for the delay.

**Comment by jywarren** [comment]
No problem, thank you!

**Comment by Souravirus** [comment]
Hi @jywarren, I have worked on the change of the navbar for the large screens and now it looks kind of decent. Here is a screenshot of it. 
screenshot from 2018-12-17 10-05-25
I know its little different than the bootstrap3 navbar. But if you want any change in the current navbar. Please tell me. Now I have to work on the navbar in medium and small screens. Thank you!!

**Comment by Souravirus** [comment]
Hi @jywarren,
I have corrected the links in navbar when the user is logged in as well as when the user is logged out for the large and medium screens. Here are some screenshots to show:
screenshot from 2018-12-20 19-40-01
screenshot from 2018-12-20 19-40-06
screenshot from 2018-12-20 19-40-17

There is some problem with the dropdown in small screens but I am sure I will find something get it working. Please see these screenshots and tell if anything is inappropriate. I have tried to my best to mimic the look of the navbar in Bootstrap 3.
Thank you!!

**Comment by jywarren** [comment]
This all looks fantastic. Keep going and great work!

**Comment by Souravirus** [comment]
Hi @jywarren, the dropdowns in bootstrap 4 are not as good as bootstrap 3. So, here are the screenshots of the dropdowns in small screens:
screenshot from 2018-12-22 00-31-40
screenshot from 2018-12-22 00-32-18
Is it ok?
Thank you!!

**Comment by Souravirus** [comment]
Hi @jywarren,
Due to the recent changes in the rich editor, the map is not visible in the bootstrap 4 version. Can you help me with this? Here is the screenshot:
screenshot from 2019-02-09 00-05-52

**Comment by jywarren** [comment]
hmm, maybe @sagarpreet-chadha has an idea?

**Comment by Souravirus** [comment]
Hi @jywarren, That map problem is not appearing on unstable.publiclab.org. So, everything is cool. You can see the upgrades on the unstable branch if you want. Some things seem incomplete. I will complete them soon.

**Comment by CoderJolly** [comment]
@Souravirus Where can I help ?

**Comment by Souravirus** [comment]
Hi @CoderJolly, thanks for showing interest in the upgrade. Ok so, first just pull this branch and just see which HTML files in app/views folder are still not covered. Now those which are not covered replace the bootstrap 3 classes with their respective bootstrap 4 classes.
Good luck !!

**Comment by Souravirus** [comment]
Hi @jywarren, I had a problem with installing will_paginate-bootstrap4 in Jenkins though it's installing in my computer. Here is the link of the installation in Jenkins https://jenkins.laboratoriopublico.org/job/Plots-Unstable/468/console. 
Thank you!!

**Comment by jywarren** [comment]
Hmm, can you share your gem sources and perhaps we need to update them somewhere? Is this available from standard RubyGems? 

```
Bundler::GemNotFound: Could not find will_paginate-bootstrap4-0.2.2 in any of the sources
```

`@icarito` may be able to help with this as well if you get stuck -- ping him or ask in the chatroom perhaps?

**Comment by jywarren** [comment]
> Due to the recent changes in the rich editor, the map is not visible in the bootstrap 4 version

Hmm, do you see what the errors are? Can you make a gist? Thanks!

**Comment by Souravirus** [comment]
Yeah sure, I will be seeing any error logs that are available but the maps were working perfectly in unstable branch.

**Comment by Souravirus** [comment]
Yeah, the gem is available in RubyGems. Here is the link: https://rubygems.org/gems/will_paginate-bootstrap4/versions/0.2.2. I will ping @icarito in the chatroom.
Thank you!!

**Comment by Souravirus** [comment]
Hi @jywarren and @icarito,
The problem with the jenkins test regarding the installing of will_paginate_bootstrap-4 is now solved. Thanks to the recent upgrade to yarn. It solved the problem.

**Comment by jywarren** [comment]
Oh, that's great!!

**Comment by Souravirus** [comment]
Hi @jywarren, can you guide me here? How to access the code inside the feature system? Because I wanted to change the bootstrap of the carousel on the home page.

**Comment by jywarren** [comment]
Ah, unfortunately it's driven from the database. We'll have to dump it from
the production site and save refactored versions until the moment we
publish the new code... Let's do this last!

**Comment by Souravirus** [comment]
Ok as you wish. Thank you!!

**Comment by Souravirus** [comment]
Hi @jywarren @sagarpreet-chadha, the map in /post location is giving this error in the unstable branch.
`Source map error: request failed with status 404
Resource URL: https://unstable.publiclab.org/assets/application-293c03e87b6999abf3d3a38d40c2841d11be21e56a585ca8302da969d91fb577.js
Source Map URL: noty.js.map[Learn More]`
I need some help here. Thank you!!

**Comment by jywarren** [comment]
Ah, sorry I missed this!!! @sagarpreet-chadha @icarito could you take a look here? Is this in the `noty.js` package?

**Comment by jywarren** [comment]
Can you paste in any more detail of the error?

**Comment by Souravirus** [comment]
Hi @jywarren, this was happening on the rich editor page where we ask for location and there is a map for location. So, when I saw the console through inspect element, the above error was coming and also the map was not being seen there.

**Comment by jywarren** [comment]
Wow, amazing rebase as usual 😮 

If you run it locally, and see that error, it should (in development mode) show the exact line number instead of the compiled JS. Maybe that'll give you an additional clue?

**Comment by jywarren** [comment]
I'm seeing this error, instead: https://gist.github.com/jywarren/4365f8ef841d9f27df3dbe38986827cb

image

**Comment by Souravirus** [comment]
Hmm I see, but when I do inspect element in my computer, this error is coming https://gist.github.com/Souravirus/82bf8e9c8961d30c1d5ca0db36f7156e
screenshot from 2019-03-01 23-59-09

**Comment by Souravirus** [comment]
The errors while running the website in development mode locally are:
https://gist.github.com/Souravirus/af74e9c22d87ce505312b52fac7fb459

**Comment by jywarren** [comment]
Ah, the red text is the error, the yellow is just a warning, i think. I believe it may be because that page's code is using Bootstrap tooltips in the PublicLab.Editor project. Has syntax for those changed? The Editor project initiates tooltips in a number of locations; will we have to make it Bootstrap 4 compatible? 

https://github.com/publiclab/PublicLab.Editor/search?utf8=%E2%9C%93&q=tooltip&type=

We can open an issue for this, can you link to the guidance for Bootstrap 4 tooltips? Sorry this took me a little while to figure out, but I bet this is the issue!

**Comment by Souravirus** [comment]
Hi @jywarren, I have removed the tooltip error by adding popper.js with the help of bootstrap.bundle.js. So, I guess that was not the error. Now also, some errors are there. Here is the screenshot. 
screenshot from 2019-03-05 23-21-34
And these errors are those which I am seeing locally.
Thank you!!

**Comment by jywarren** [comment]
Ah, i think my comment maybe didn't get processed by GitHub?

> OK! i think we use this lib for typeahead, so see if it needs to be updated or reinstalled? https://github.com/bassjobsen/Bootstrap-3-Typeahead

**Comment by Souravirus** [comment]
Yeah, but they haven't updated it for Bootstrap 4. So, we have to find a different workaround.

**Comment by Souravirus** [comment]
Hi @jywarren, Everything is working fine now after adding the typeahead.js to the javascript assets. 
Thanks!!

**Comment by jywarren** [comment]
Awesome! Wait, EVERYTHING?

**Comment by Souravirus** [comment]
Oh, I mean there were problems with the map in the user profile page also which was also solved.

**Comment by jywarren** [comment]
Haha awesome. Keep up the good work!

**Comment by Souravirus** [comment]
Hi @jywarren, all the views are now covered expect the carousel one which is in feature system. So, I guess its time to correct the carousel of the homepage. If you want to see the changes, you can see it in the unstable branch. I have currently pushed the changes and if there are some pages which seems odd or are not according to bootstrap 4, please tell me. I will correct them.
Thank you!!

**Comment by jywarren** [comment]
Ah i guess i see somehting on the dashboard -- small though!

image

**Comment by Souravirus** [comment]
@jywarren its currently pushed by gaurav. I will push the changes little later

**Comment by Souravirus** [comment]
Hi @jywarren, just see the changes now in the unstable branch. Thank you!!

**Comment by gauravano** [comment]
- [ ] button which shows prev versions not working route: /post  
bug1

**Comment by Souravirus** [comment]
Thanks @gauravano for reporting the error. Working on this.

**Comment by gauravano** [comment]
I think we can solve minor issues after merging this too as it can be difficult and might not be justified to lay it all on @Souravirus? 

We can ask contributors to raise fix against this PR but they might get confused too so it's better to check all the major functionalities and UI part which is necessary?

Really awesome work @Souravirus :100:  :tada: :muscle:

Literally, :raised_hands: to the dedication, patience, and work :point_down: 

Screenshot from 2019-03-19 23-20-02

**Comment by gauravano** [comment]
Also, let's tag @publiclab/community-reps as they can help us in testing in the work.

**Comment by Souravirus** [comment]
Thanks @gauravano

**Comment by jywarren** [comment]
Yes, this is a good idea @gauravano -- how might we structure this? Could
we ask people to open PRs against this PR? Or are we close enough that we
could merge it to stable, then ask people to go in and file fixes before
publishing it to the live site? Ideas?

And YES this is AMAZING!!!!! It was a gargantuan task!!!!!!! Sourav, really
amazing.

**Comment by Souravirus** [comment]
In my opinion, since this a big PR, we can first correct those errors or glitches which are seen on unstable and then we can push them to the stable and then to live site. What do you think about this @gauravano @jywarren?

**Comment by gauravano** [comment]
If you feel this PR still need more eyes @Souravirus. We can give this PR a week or two on unstable.publiclab.org and can ask all the contributors to test and raise PR against your branch. So, would it be easy for contributors to raise PR against a plots2 branch? If yes, could you push this branch to plots2 Sourav?

**Comment by jywarren** [comment]
Just adding a few; maybe we can do a first pass of major ones, then go to the stable workflow?

* [ ] Header folding at narrower widths (see image below)

image

* [ ] tags in "label" style appearing as plain text - see image below:

image

**Comment by Souravirus** [comment]
@gauravano, the suggestions given and the chats on this PR will be lost. So, I guess those who want to contribute can push directly to this branch as there is permission to push commits to this branch. What do you think?

**Comment by gauravano** [comment]
So, let's create an issue where we would mention all the steps(in detail) that how a contributor can help us in fixing a bug at unstable.publiclab.org? How does it sound?

**Comment by Souravirus** [comment]
Yeah, that would be really great.

**Comment by Souravirus** [comment]
Hi @jywarren regarding the folding up the header, I have removed the default caret down symbol from the dropdown and now in the narrow screens, it's looking like this.
Screenshot from 2019-03-20 01-16-16
Is this appropriate?

**Comment by Souravirus** [comment]
@gauravano, if you want to add anything more to the issue, please add.
Thank you!!

**Comment by jywarren** [comment]
Hmm, shall we make the search box a bit smaller then too? Oh, actually in
bootstrap 3, all menu items are collapsed under the ||| menu... can we do
that here too?

**Comment by Souravirus** [comment]
Hi @jywarren, the menu items do collapse here, but only for extra small and small screens. If you want the collapsing to be expanded till medium screen, I can do that. Also, trying to make the search box a little smaller.

**Comment by Souravirus** [comment]
Hi @jywarren, the folding of the navbar is now solved. Here are some screenshots of the medium screen. 
Screenshot from 2019-03-21 20-26-24

**Comment by Souravirus** [comment]
The tag error is also solved as I haven't replaced badge-default (labels are replaced by badges in bootstrap 4) to badge-secondary. Now it's also solved. Here is the screenshot
image
Thank you!!

**Comment by jywarren** [comment]
Great - it's odd, can we get the tag 'x' to be white? Thanks for this and
the header fix!

**Comment by Souravirus** [comment]
Hi @jywarren, so the 'x' button is now changed to white. Here is the screenshot
image

**Comment by jywarren** [comment]
Super!

**Comment by Souravirus** [comment]
Hi @jywarren, @gauravano there are some errors with the javascript on the rich editor page. That's why that more button which @gauravano told earlier was not working. Here is the picture of the errors.
image
The error with the editor not defined is due to the checkbox calling the editor variable here https://github.com/publiclab/plots2/blob/64b02071dc8f04f99e3d60fb41ee4d426c658e0b/app/views/editor/rich.html.erb#L380-L388.
 I know that the editor variable is already declared there in this line https://github.com/publiclab/plots2/blob/64b02071dc8f04f99e3d60fb41ee4d426c658e0b/app/views/editor/rich.html.erb#L263.
Please help me here.

**Comment by Souravirus** [comment]
I guess this is happening due to that first error. Somewhere the PublicLab editor is not syncing with Bootstrap 4. I am into it though. Let's see

**Comment by Souravirus** [comment]
https://github.com/publiclab/PublicLab.Editor/blob/b1436aa71facade76f0a878e46b0eac7c003c208/dist/PublicLab.Editor.js#L21744 . Please see this line. I guess here is some error. Also, PublicLab editor is in Bootstrap 3. So, should that be of any problem?

**Comment by jywarren** [comment]
I hope that enough of the classes are the same that if we change the HTML on plots2, the editor won't mind, but we will have to see! 

Can you try instantiating an editor manually in the JS console, like `ed = new PublicLab.Editor()` to see what happens?

**Comment by Souravirus** [comment]
Hi @jywarren, this is the error that's coming.
image

**Comment by Souravirus** [comment]
But I think this is not a problem as this error is coming with the previous version of Bootstrap also.

**Comment by jywarren** [comment]
Ah, but i think you need to pass some options in your test statement, so
like ed= new PublicLab.Editor({...}) -- try with just what's on the page
already?

**Comment by Souravirus** [comment]
Hi @jywarren, this is the error that is coming while pasting the options on the page list:
image

**Comment by jywarren** [comment]
Ah! It looks like it's creating a Leaflet map twice. @sagarpreet-chadha,
any idea here?

**Comment by jywarren** [comment]
Hmm, i can't see any duplicated Leaflet includes or anything. Can you look at the Network pane in Chrome dev tools and see if somehow the Leaflet library is getting dual-loaded? in development mode you should be able to see each JS file loaded separately.

**Comment by Souravirus** [comment]
Oh OK Thanks @jywarren for seeing this out. Yeah I will check it out in chrome Dev tools. 
Thank you!!

**Comment by Souravirus** [comment]
@jywarren, Actually there might be some error in my command because its occurring in chrome browser also. Don't you think something is wrong in this line. 
https://github.com/publiclab/PublicLab.Editor/blob/d57797d3bfc2fe13b643e87cbfb574ee027f8b97/dist/PublicLab.Editor.js#L21744

**Comment by Souravirus** [comment]
https://gist.github.com/Souravirus/dbc9bb14ddd8f83cd9e1d48e7b06cbd2 Here is the gist about the command I typed in web console.

**Comment by jywarren** [comment]
Hmm... I wonder if @rexagod can suggest anything about this? https://github.com/publiclab/PublicLab.Editor/blob/e4fcc6b769f3ad81364237924def5bc97db4f015/src/modules/PublicLab.Module.js#L13

Do we need to initialize `options` as `{}` first?

**Comment by jywarren** [comment]
Oh! but sorry, rexagod is in exams!

**Comment by jywarren** [comment]
Wait! Is it related to this? https://github.com/publiclab/PublicLab.Editor/issues/299

**Comment by jywarren** [comment]
Hmm. Being related to guides, i wonder about the guides code here:

https://github.com/publiclab/PublicLab.Editor/blob/e4fcc6b769f3ad81364237924def5bc97db4f015/src/modules/PublicLab.Module.js#L19-L36

**Comment by jywarren** [comment]
Ah wait, so re `Hi @jywarren, this is the error that is coming while pasting the options on the page list:` that's because we have already started up the editor and it had a map. You could set `mapModule: false` but not sure quite what will happen then.

**Comment by jywarren** [comment]
I mean in your testing, not in the actual PR.

**Comment by jywarren** [comment]
I wonder if this is a timing issue. Could you wrap the definition for `draftCheck` in a closure, so it loads only after the page is done loading? 

https://github.com/publiclab/plots2/blob/8b57e841df2efe7d7545f4ae516dee8e9704e330/app/views/editor/rich.html.erb#L375

so:

```js
var draftCheck;
(function() {

  draftCheck = function draftCheck() {

  }
})();
```

I think this will work.

**Comment by Souravirus** [comment]
Hi @jywarren, I will be testing this out soon.
Thank you!!

**Comment by Souravirus** [comment]
This timing issue is also not there. I tried but it is also not working.

**Comment by Souravirus** [comment]
Hi @jywarren, So, I tried to make the `mapModule:false` and every javascript that is at the footer is working. But the thing is that the map now doesn't appear. So, what should I do?

**Comment by jywarren** [comment]
Let's move forward and I can try debugging this later. Can you make a
checklist item for before this gets merged? Thanks!!!!

**Comment by Souravirus** [comment]
Hi @jywarren, besides that I think everything is working fine. But there was one thing the carousel at the home page which you said will dump from production. So, we have to see that only. Else everything is sorted. I will just put the new code in the `unstable` branch just to see everything is working fine. 
Thank you!!

**Comment by Souravirus** [comment]
There was one test failing which got corrected and some glitches in the site which got cleared. You can check the updated site at the unstable and I think it's ready to merge now.

**Comment by rexagod** [comment]
Hello @Souravirus! Amazing work! I just wanted to chime in here to check if there's anything in the Editor, or JS generally, that you may need any help with? I'd be happy to help if you could point me in a direction! :smile:

Thanks and again, great work! :+1:

**Comment by Souravirus** [comment]
Yeah hi @rexagod, just wanted to say there is this error coming in the rich editor `_module.options.guides is undefined` and its pointing to [here](https://github.com/publiclab/PublicLab.Editor/blob/d57797d3bfc2fe13b643e87cbfb574ee027f8b97/dist/PublicLab.Editor.js#L21759). So, please see to it. 
Thank you !!

**Comment by rexagod** [comment]
@Souravirus, can you guide me through the exact steps that can reproduce this error? Thanks!

**Comment by Souravirus** [comment]
Hi @rexagod, you can see it through the browser console and just navigate to the root cause of the error and you will find it.

**Comment by rexagod** [comment]
Seems fixed now. @Souravirus please check on your local as well. Thanks!

Screenshot from 2019-04-10 18-14-06

**Comment by Souravirus** [comment]
Hey @rexagod, thanks it worked. Nice work!!

File: package.json [hunk 3]
```diff
@@ -50,7 +50,7 @@
     "noty": "3.1.4",
     "phantomjs-prebuilt": "^2.1.16",
     "popper.js": "^1.4.0",
-    "publiclab-editor": "^2.0.0",
+    "publiclab-editor": "~2.0.0",
```
Comment (rexagod) [inline package.json]: /cc @jywarren

**Comment by Souravirus** [comment]
Hi @jywarren, when are we merging this?

**Comment by jywarren** [comment]
Hi, @Souravirus if you think we are ready and have checked enough, let's do it, and we can keep bugfixing on stable if we think it's an issue. Thank you so much. I'll look out tomorrow to merge this, heads up @publiclab/reviewers @publiclab/plots2-reviewers @gauravano @SidharthBansal @cesswairimu @sashadev-sky @milaaraujo @stefannibrasil @CleverFool77 @gautamig54 @divyabaid16 @namangupta01 @ViditChitkara @siaw23 and anyone else who's been working on anything related to UI!!!

**Comment by jywarren** [comment]
Get ready for the big day!!!!!!!

**Comment by jywarren** [comment]
Feeling good here @Souravirus? Shall I click MERGE?

**Comment by Souravirus** [comment]
Ok just one thing when should we correct the carousel thing as you told to dump that from the production.

**Comment by jywarren** [comment]
Yes, i'll share a gist right now. We'll have time before it goes to production. Otherwise shall I merge?

**Comment by jywarren** [comment]
https://gist.github.com/jywarren/eb209b174c223ca89991addf10d3102a

**Comment by Souravirus** [comment]
Yeah its fine to merge. EXCITED!!

**Comment by jywarren** [comment]
THATS IT!!!!! 🙌🙌🙌🙌 ⚡️⚡️⚡️⚡️

**Comment by jywarren** [comment]
i'tll build to stable now...

**Comment by jywarren** [comment]
if u want to share a new version of the feature code with me i can put it up on stable for testing.

**Comment by jywarren** [comment]
i mean the carousel

**Comment by Souravirus** [comment]
Ok, I will share the updated code of the carousel in a gist in some time. 
Thank you!!
Atlast it is merged. :smile:

**Pull request merged**